### PR TITLE
Release 21.3.0

### PR DESCRIPTION
**Pull request by altendky**

### CODE REVIEW

**Comment by altendky** [comment]
Ready for the first review for tagging and publishing 21.3.0rc1.

Release email draft...

---

towncrier releases 19.9.0 and 21.3.0rc1

There are a couple towncrier releases...  in process.  This is a bit more complicated than normal since 19.9.0rc1 was never followed by a non-RC release.  You can see a detailed checklist and discussion of this in #313 [0].  19.9.0 artifacts have been created [1] and are ready to be published, but are being held.  Relative to 19.9.0rc1, 19.9.0 contains a few readme fixes and the news fragments processed into news file.  21.3.0rc1 has been published and is open for a week for feedback before 21.3.0 is planned to be released.  19.9.0 publishing is being held until after 21.3.0 to avoid a somewhat pointless intermediate step for developers that are using 'the latest non-pre release on PyPI' (presently 19.2.0).  No real need for someone to get 19.9.0 now just to see 21.3.0 available next week.  The news for 21.3.0 is available for review [2].

There are remaining bugs and important features.  But, it took long enough to get to this point that I consider it important to get something out.  Several people are waiting patiently [3].  As we find time to design and address the remaining issues, hopefully we can manage more frequent releases.

Please do let me know of any issues you find that are not already documented.  Or, if any of the documented ones should really be release blockers.  I sure hope we can get this out though.  :]

Cheers,
-kyle


[0]: https://github.com/twisted/towncrier/issues/313
[1]: https://github.com/twisted/towncrier/pull/331#issuecomment-803491163
[2]: https://github.com/twisted/towncrier/blob/21.3.0rc1/NEWS.rst
[3]: https://github.com/twisted/towncrier/issues/269

File: NEWS.rst
Comment (altendky) [inline NEWS.rst:46]: Hmm...  Why did I think that misc was ignored...  bah.

Comment (altendky) [reply NEWS.rst:46]: https://github.com/twisted/towncrier/blob/4143e0ff55177cca5747d18624d614e593ebe3c3/pyproject.toml#L30-L33

Is this supposed to do what's above?  Or totally not include it in the output?  Guess I'll dive into the tests to try to understand.

Comment (altendky) [reply NEWS.rst:46]: It seems that misc hasn't been used since [17.4.0](https://github.com/twisted/towncrier/blob/4143e0ff55177cca5747d18624d614e593ebe3c3/NEWS.rst#misc-1) but it does look quite intentionally formatted.  I guess this is good.

image

Comment (adiroiban) [reply NEWS.rst:46]: This is how towncrier works by default... and how it's used in Twisted (without the extra links).

It only hides the "content" and not the note number itself :p

I am using a custom template in which `showcontent` is completely ignored.

Maybe besides `showcontent` the default template can have a new `ignore` that can be used to now show that type at all in the release notes.

**Comment by adiroiban** [comment]
My suggestion for the email

-------------

Subject: towncrier releases 19.9.0 and 21.3.0rc1

Hi everyone,

I am announcing 2 twisted releases.

This is a bit more complicated than normal since 19.9.0rc1 was never followed by a non-RC release.  

You can see a detailed checklist and discussion of this in #313 [0]. 

Relative to 19.9.0rc1, 19.9.0 contains a few readme fixes and the news fragments processed into news file. 

19.9.0 artifacts have been created [1] and 19.9.0 publishing is being held until after 21.3.0 to avoid a somewhat pointless intermediate step for developers that are using 'the latest non-pre release on PyPI' (presently 19.2.0).
 No real need for someone to get 19.9.0 now just to see 21.3.0 available next week.  The news for 21.3.0 is available for review [2].

21.3.0rc1 has been published and is open for a week for feedback before 21.3.0 is planned to be released. 

There are remaining bugs and important features.  But, it took long enough to get to this point that I consider it important to get something out.  Several people are waiting patiently [3].  As we find time to design and address the remaining issues, hopefully we can manage more frequent releases.

Please do let me know of any issues you find that are not already documented.  Or, if any of the documented ones should really be release blockers.  I sure hope we can get this out though.  :]

Cheers,
-kyle


[0]: https://github.com/twisted/towncrier/issues/313
[1]: https://github.com/twisted/towncrier/pull/331#issuecomment-803491163
[2]: https://github.com/twisted/towncrier/blob/21.3.0rc1/NEWS.rst
[3]: https://github.com/twisted/towncrier/issues/269

File: NEWS.rst
Comment (adiroiban) [reply NEWS.rst:46]: Minor comment. Can we manually remove  the misc section?

**Review by adiroiban — APPROVED** [review]
All good. Thanks!

**Comment by altendky** [comment]
So, that's cool.  `pep517.meta` seems to normalize `dev4` and `rc1` differently in regards to a `.` or not...

```console
$ venv/bin/python -c 'import pep517.meta; print(pep517.meta.load(".").version)'
WARNING: You are using pip version 20.3.3; however, version 21.0.1 is available.
You should consider upgrading via the '/home/altendky/repos/towncrier/venv/bin/python -m pip install --upgrade pip' command.
WARNING: You are using pip version 20.3.3; however, version 21.0.1 is available.
You should consider upgrading via the '/home/altendky/repos/towncrier/venv/bin/python -m pip install --upgrade pip' command.
21.3.0rc1
```

```console
$ venv/bin/python -c 'import pep517.meta; print(pep517.meta.load(".").version)'
WARNING: You are using pip version 20.3.3; however, version 21.0.1 is available.
You should consider upgrading via the '/home/altendky/repos/towncrier/venv/bin/python -m pip install --upgrade pip' command.
WARNING: You are using pip version 20.3.3; however, version 21.0.1 is available.
You should consider upgrading via the '/home/altendky/repos/towncrier/venv/bin/python -m pip install --upgrade pip' command.
21.3.0.dev4
```

```console
$ venv/bin/python -c 'import incremental; versions = [incremental.Version("", 21, 3, 0, dev=4), incremental.Version("", 21, 3, 0, release_candidate=1)]; print("\n".join(version.public() for version in versions))'
21.3.0.dev4
21.3.0.rc1
```

And thus we get...

https://github.com/twisted/towncrier/runs/2198852633?check_suite_focus=true
image

Just manually upload the CI artifacts to PyPI?  Delete the tag, choose how to fix this, go again?  Just push tag `21.3.0rc1`?  Maybe this comparison should use `packaging.utils.canonicalize_version()` so it is less sensitive.

@graingert, @adiroiban

**Comment by adiroiban** [comment]
> Just manually upload the CI artifacts to PyPI? Delete the tag, choose how to fix this, go again? Just push tag `21.3.0rc1`? Maybe this comparison should use `packaging.utils.canonicalize_version()` so it is less sensitive.

I think you can do a manual upload.... and we can look at updating incremental or more probably update "admin/check_tag_version_match.py" to use the same incremental version

The normalization done by pep517 module is the one described in PEP440.

Thanks

**Comment by altendky** [comment]
The weird bit about `pep517.meta`/pip is that it outputs `21.3.0.dev4` (with the `.`) but also `21.3.0rc1` (without the `.`).  Which is why it worked for the test, using `.dev4`, but failed here with `.rc1`.

https://www.python.org/dev/peps/pep-0440/#pre-release-separators
> Pre-releases should allow a ., -, or _ separator between the release segment and the pre-release segment. The normal form for this is without a separator.

`pep517.meta.load().version` doesn't, by way of its name, exactly claim to normalize.  But, it is a bit odd that it does, partially, in one case and doesn't also apply the same change not in another case.

**Comment by altendky** [comment]
Oh huh, that's fun...  Since we build on-tag and we tag on-the-PR-branch and we build on-PR...  we double up here.

image

**Comment by adiroiban** [comment]
> Oh huh, that's fun... Since we build on-tag and we tag on-the-PR-branch and we build on-PR... we double up here.

Which build is this?

-------------

I see `CI / Check Newsfragment` failing on the release branch and I think there is already a reported issue and maybe a PR...
So I think that we should prioritize this bug.

I think that adding a newsfragment as in bd963c3  should not be part of a normal release process

File: NEWS.rst
Comment (adiroiban) [inline NEWS.rst:8]: Very minor comment. Feel free to ignore.

Maybe be explicit there are no changes, other than version update.

One might think there might be some change and try to search for them...

```suggestion
No significant changes since the previous release candidate.
```

**Review by adiroiban — APPROVED** [review]
All good. Thanks. Ready for  final release  :)

Maybe future work.

* Fix towncrier release check to not complain about missing news fragment for a release branch.... the logic should be reversed for a release and complain if there are newsfragments
* Maybe add a help CLI to generate a final release section with current date and time and maybe with a body like `No changes since the previous release.`

**Comment by altendky** [comment]
> * Fix towncrier release check to not complain about missing news fragment for a release branch.... the logic should be reversed for a release and complain if there are newsfragments

I think that's https://github.com/twisted/towncrier/issues/309.  Or at least closely related.

> * Maybe add a help CLI to generate a final release section with current date and time and maybe with a body like `No changes since the previous release.`

This sounds like a bigger suggestion but all I see as different than present functionality is the addition of time to the template and the change in wording of the 'nothing to see here' comment when there are no newsfragments.

**Comment by adiroiban** [comment]
Thanks for the ticket link.
The CLI was a suggestion for a separate future feature... not sure if that is common.

**Comment by altendky** [comment]
I'm not clear in what scenario we would need a separate CLI (or subcommand or...) for doing that.  Presumably I don't understand the feature yet.  Feel free to pick whatever communications option you see fit to help me understand.  `:]`

**Review by adiroiban — APPROVED** [review]
All good. Thanks!

I don't think that a review is required for the master merge. But here we go :)

**Comment by altendky** [comment]
Welp, you failed at the review, sorry.  I haven't updated the rev to be not-21.3.0 yet.  I forgot to set it to draft before dismissing the review.  This review/dismiss/review/...  is not great.  Haven't thought about how to change it though, other than \_more\_ PRs.

**Comment by altendky** [comment]
Oh, also...  ["Merge without waiting for an approving review."](https://github.com/twisted/towncrier/blob/19eb09bcf3b1c590248024975d812aeff9110f12/CONTRIBUTING.rst)  `:]`

**Comment by adiroiban** [comment]
> Welp, you failed at the review, sorry.

True. My bad.

**Comment by altendky** [comment]
Should have put a smiley on that.  Hopefully you can accept this backwards one retroactively.  `[:`

**Pull request merged**
